Bearded.Utilities/Noise/PerlinNoise.cs

```diff
@@ -0,0 +1,113 @@
+            .Select(vector => new Vector2d(vector.X, vector.Y))
+            .ToArray();
+
+        /// <summary>
+        /// Generates a procedural texture using Perlin noise.
+        /// </summary>
+        /// <remarks>
+        /// Perlin noise is generated by first generating a grid of gradient vectors. The texture is then filled in by
+        /// for each point identifying the four surrounding gradient vectors, calculating the dot product between each
+        /// of the gradient vectors and corresponding offset vector, and interpolating between them.
+        /// </remarks>
+        /// <param name="numCellsX">
+        /// The number of cells in which the texture is divided along the horizontal axis (see remarks).
+        /// </param>
+        /// <param name="numCellsY">
+        /// The number of cells in which the texture is divided along the vertical axis (see remarks).
+        /// </param>
+        /// <param name="seed"></param>
+        /// <returns></returns>
+        /// <exception cref="ArgumentOutOfRangeException"></exception>
+        public static IProceduralTexture Generate(int numCellsX, int numCellsY, int? seed)
+        {
+            if (numCellsX <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(numCellsX));
+            }
+            if (numCellsY <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(numCellsY));
+            }
+
+            var gradientArray = NoiseUtils.GenerateRandomArray(
+                // We generate the corners, but to make the noise map wrap along both dimensions, we reuse the same
+                // values as the left and top boundaries, so we don't have to generate the right and bottom boundaries.
+                numCellsX,
+                numCellsY,
+                r => vectorSamples[r.Next(vectorSamples.Length)],
+                seed);
+            return new PerlinProceduralTexture(gradientArray);
+        }
+
+        private sealed class PerlinProceduralTexture : IProceduralTexture
+        {
+            private readonly Vector2d[,] gradientArray;
+            private readonly int width;
+            private readonly int height;
+
+            public PerlinProceduralTexture(Vector2d[,] gradientArray)
+            {
+                this.gradientArray = gradientArray;
+                width = gradientArray.GetLength(0);
+                height = gradientArray.GetLength(1);
+            }
+
+            public double ValueAt(double x, double y)
+            {
+                if (x < 0 || x >= 1)
+                {
+                    throw new ArgumentOutOfRangeException(nameof(x));
+                }
+                if (y < 0 || y >= 1)
+                {
+                    throw new ArgumentOutOfRangeException(nameof(y));
+                }
+
+                // First we transform x and y from [0, 1) x [0, 1) to [0, width) x [0, height) for easier math.
```

Comment:
Removed the comments and replaced it with more methods. Not sure if this is clearer, I think I prefer the comments, but unless you tell me otherwise, I'll leave it in this new form.